OrbitLinuxTracing/TracerThread.cpp

```diff
@@ -89,98 +107,122 @@ void TracerThread::InitUprobesEventProcessor() {
+  return true;
+}
+
+void TracerThread::AddUprobesFileDescriptors(
+    const absl::flat_hash_map<int32_t, int>& uprobes_fds_per_cpu,
+    const LinuxTracing::Function& function) {
+  for (const auto [cpu, fd] : uprobes_fds_per_cpu) {
+    uint64_t stream_id = perf_event_get_id(fd);
+    uprobes_uretprobes_ids_to_function_.emplace(stream_id, &function);
+    uprobes_ids_.insert(stream_id);
+    tracing_fds_.push_back(fd);
+    fds_per_cpu_[cpu].push_back(fd);
+  }
+}
+
+void TracerThread::AddUretprobesFileDescriptors(
+    const absl::flat_hash_map<int32_t, int>& uretprobes_fds_per_cpu,
+    const LinuxTracing::Function& function) {
+  for (const auto [cpu, fd] : uretprobes_fds_per_cpu) {
+    uint64_t stream_id = perf_event_get_id(fd);
+    uprobes_uretprobes_ids_to_function_.emplace(stream_id, &function);
+    uretprobes_ids_.insert(stream_id);
+    tracing_fds_.push_back(fd);
+    fds_per_cpu_[cpu].push_back(fd);
+  }
+}
+
+bool TracerThread::OpenUserSpaceProbes(const std::vector<int32_t>& cpus) {
   bool uprobes_event_open_errors = false;
-  absl::flat_hash_map<int32_t, int> uprobes_ring_buffer_fds_per_cpu;
 
   for (const auto& function : instrumented_functions_) {
-    absl::flat_hash_map<int32_t, int> function_uprobes_fds_per_cpu;
-    absl::flat_hash_map<int32_t, int> function_uretprobes_fds_per_cpu;
-    bool function_uprobes_open_error = false;
-
-    for (int32_t cpu : cpus) {
-      int uprobes_fd = uprobes_retaddr_event_open(
-          function.BinaryPath().c_str(), function.FileOffset(), -1, cpu);
-      if (uprobes_fd < 0) {
-        function_uprobes_open_error = true;
-        break;
+    absl::flat_hash_map<int32_t, int> uprobes_fds_per_cpu;
+    absl::flat_hash_map<int32_t, int> uretprobes_fds_per_cpu;
+    uint64_t address = function.VirtualAddress();
+
+    if (manual_instrumentation_config_.IsTimerStartAddress(address)) {
+      // Only open uprobe for a "timer start" manual instrumentation function.
+      if (!OpenUprobes(function, cpus, &uprobes_fds_per_cpu)) {
+        CloseFileDescriptors(uprobes_fds_per_cpu);
+        uprobes_event_open_errors = true;
+        continue;
       }
-      function_uprobes_fds_per_cpu.emplace(cpu, uprobes_fd);
-
-      int uretprobes_fd = uretprobes_event_open(function.BinaryPath().c_str(),
-                                                function.FileOffset(), -1, cpu);
-      if (uretprobes_fd < 0) {
-        function_uprobes_open_error = true;
-        break;
-      }
-      function_uretprobes_fds_per_cpu.emplace(cpu, uretprobes_fd);
-    }
-
-    if (function_uprobes_open_error) {
-      ERROR("Opening u(ret)probes for function at %#016lx",
-            function.VirtualAddress());
-      uprobes_event_open_errors = true;
-      for (const auto& uprobes_fd : function_uprobes_fds_per_cpu) {
-        close(uprobes_fd.second);
+    } else if (manual_instrumentation_config_.IsTimerStopAddress(address)) {
+      // Only open uretprobe for a "timer stop" manual instrumentation function.
```

Comment:
I guess we can go with uretprobes for now but I would still be curious to get an answer on this (also I guess it would be interesting to measure).

Reply:
Ah sorry for I missed the question. It has to do with the duplicate uprobe detection code that relies on the presence of a uretprobe. Also, it's easier to have one uprobe and one uretprobe so that we automatically have the proper stacking of timeslice events on the collector side.

Reply:
Ah right, so that all the other processing is done in exactly the same way as dynamic instrumentation.